pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -27,65 +27,133 @@ import (
-	// will be compatible with this node
+
 	for _, p := range pods {
 		n.Add(p)
 	}
-	return n
+	if len(n.InstanceTypeOptions) == 0 {
+		return nil, errors.New("no instance type satisfied requirements")
+	}
+	return n, nil
 }
 
 func (n Node) Compatible(pod *v1.Pod) error {
-	if n.unschedulable {
-		return errors.New("node is unschedulable")
-	}
-
 	podRequirements := v1alpha5.NewPodRequirements(pod)
-	if err := n.Constraints.Requirements.Compatible(podRequirements); err != nil {
+	if err := n.Requirements.Compatible(podRequirements); err != nil {
 		return err
 	}
 
-	tightened := n.Constraints.Requirements.Add(podRequirements.Requirements...)
+	tightened := n.Requirements.Add(podRequirements.Requirements...)
 	// Ensure that at least one instance type of the instance types that we are already narrowed down to based on the
 	// existing pods can support the pod resources and combined pod + provider requirements
+	podRequests := resources.RequestsForPods(pod)
 	for _, it := range n.InstanceTypeOptions {
-		if cloudprovider.Compatible(it, tightened) && n.hasCompatibleResources(resources.RequestsForPods(pod), it) {
+		newSize := resources.Merge(n.reservedResources(it), podRequests)
+		if cloudprovider.Compatible(it, tightened) &&
+			n.newPodCanFit(newSize, it) &&
+			n.hasCompatibleResources(resources.RequestsForPods(pod), it) {
 			return nil
 		}
 	}
 	return errors.New("no matching instance type found")
 }
 
+func (n Node) reservedResources(it cloudprovider.InstanceType) v1.ResourceList {
+	return resources.Merge(it.Overhead(), n.daemonResources, n.podResources)
+}
+
+func (n *Node) newPodCanFit(newSize v1.ResourceList, it cloudprovider.InstanceType) bool {
+	for resourceName, totalQuantity := range it.Resources() {
+		reservedQuantity := newSize[resourceName]
+		if reservedQuantity.Cmp(totalQuantity) > 0 {
+			return false
+		}
+	}
+
+	instancePodMax := it.Resources()[v1.ResourcePods]
+	if !instancePodMax.IsZero() && instancePodMax.CmpInt64(int64(len(n.Pods)+1)) < 0 {
+		return false
+	}
+	return true
+}
+
+func (n *Node) RequiredResources() v1.ResourceList {
+	if len(n.InstanceTypeOptions) == 0 {
+		return resources.Merge(n.daemonResources, n.podResources)
+	}
+	return resources.Merge(n.daemonResources, n.InstanceTypeOptions[0].Overhead(), n.podResources)
+}
+
 // Add adds a pod to the Node which tightens constraints, possibly reducing the available instance type options for this
 // node
 func (n *Node) Add(pod *v1.Pod) {
-	n.Pods = append(n.Pods, pod)
-	n.Constraints = n.Constraints.Tighten(pod)
+	n.Requirements = n.Requirements.Add(v1alpha5.NewPodRequirements(pod).Requirements...)
+	n.recalculateDaemonResources()
+
+	podRequests := resources.RequestsForPods(pod)
 	var instanceTypeOptions []cloudprovider.InstanceType
 	for _, it := range n.InstanceTypeOptions {
-		if cloudprovider.Compatible(it, n.Constraints.Requirements) &&
+		newSize := resources.Merge(n.reservedResources(it), podRequests)
```

Comment:
This calculation is unnecessary if line 146 is false. Does it make sense to compute this inside of `n.newPodCanFit(pod, instanceType)`